zarr/codecs.py

```diff
@@ -485,6 +485,105 @@ def __repr__(self):
 codec_registry[Delta.codec_id] = Delta
 
 
+class AsType(Codec):
+    """Filter to convert data between different types.
+
+    Parameters
+    ----------
+    dtype : dtype
+        Data type used to store the data.
+    astype : dtype, optional
+        Data type of the data read in.
+
+    Notes
+    -----
+    If `astype` is an integer data type, please ensure that it is
+    sufficiently large to store encoded values. No checks are made and data
+    may become corrupted due to integer overflow if `astype` is too small.
+    Note also that the encoded data for each chunk includes the absolute
+    value of the first element in the chunk, and so the encoded data type in
+    general needs to be large enough to store absolute values from the array.
+
+    Examples
+    --------
+    >>> import zarr
+    >>> import numpy as np
+    >>> x = np.arange(100, 120, 2, dtype='i1')
+    >>> x
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118], dtype=int8)
+    >>> f = zarr.AsType(dtype=x.dtype, astype='i8')
+    >>> y = f.decode(x)
+    >>> y
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118])
+    >>> z = f.encode(y)
+    >>> z
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118], dtype=int8)
+
+    """  # flake8: noqa
+
+    codec_id = 'astype'
+
+    def __init__(self, dtype, astype=None):
```

Comment:
So at least initially I like the idea of `casting` as an argument. However, my concern is this makes conversion asymmetric, but in a way that might not be obvious.

For example, say we have some `int` initially and we want to use `astype` to convert it to some `float`. No matter what conversion I chose the conversion from `int` to `float` works fine. However, if one tries to set a value in the view, this won't work unless the casting is `unsafe`.

We can come up with other types where a different casting constraint sets this threshold of whether it proceeds or not. Though instead of getting an unclear error about type conversion, maybe the better parameter instead of `casting` is simply whether we make the view `read_only` or not.

What do you think?